File: infra/src/main/java/com/beat/infra/persistence/user/repository/UsersRepositoryImpl.java
package com.beat.infra.persistence.user.repository;

import java.util.List;
import java.util.Optional;

import org.springframework.stereotype.Repository;

import com.beat.domain.user.domain.Users;
import com.beat.domain.user.repository.UserRepository;
import com.beat.infra.persistence.user.entity.UsersJpaEntity;
import com.beat.infra.persistence.user.mapper.UsersPersistenceMapper;

import lombok.RequiredArgsConstructor;

@Repository
@RequiredArgsConstructor
public class UsersRepositoryImpl implements UserRepository {

	private final UsersJpaRepository usersJpaRepository;
	private final UsersPersistenceMapper usersPersistenceMapper;

	@Override
	public Optional<Users> findById(Long id) {
		return usersJpaRepository.findById(id)
			.map(usersPersistenceMapper::toDomain);
	}

	@Override
	public List<Users> findAll() {
		return usersJpaRepository.findAll().stream()
			.map(usersPersistenceMapper::toDomain)
			.toList();
	}

	@Override
	public Users save(Users users) {
		UsersJpaEntity entity = usersPersistenceMapper.toEntity(users);
		UsersJpaEntity savedEntity = usersJpaRepository.save(entity);
		return usersPersistenceMapper.toDomain(savedEntity);
	}

	@Override
	public void delete(Users users) {
		if (users.getId() != null) {
			usersJpaRepository.deleteById(users.getId());
		}
	}
}

File: infra/src/main/kotlin/com/beat/infra/persistence/user/repository/UsersRepositoryImpl.kt
package com.beat.infra.persistence.user.repository

import com.beat.domain.user.model.Users
import com.beat.domain.user.repository.UserRepository
import com.beat.infra.persistence.user.mapper.UsersPersistenceMapper
import org.springframework.stereotype.Repository
import java.util.Optional

@Repository
class UsersRepositoryImpl(
    private val usersJpaRepository: UsersJpaRepository,
    private val usersPersistenceMapper: UsersPersistenceMapper,
) : UserRepository {
    override fun findById(id: Long?): Optional<Users> =
        usersJpaRepository.findById(requireNotNull(id) { "The given id must not be null" })
            .map(usersPersistenceMapper::toDomain)

    override fun findAll(): List<Users> =
        usersJpaRepository.findAll().map(usersPersistenceMapper::toDomain)

    override fun save(users: Users): Users =
        usersPersistenceMapper.toDomain(usersJpaRepository.save(usersPersistenceMapper.toEntity(users)))

    override fun delete(users: Users) {
        users.getId()?.let(usersJpaRepository::deleteById)
    }
}
